packages/backend/src/services/ProjectService/ProjectService.ts

```diff
@@ -1885,6 +1896,114 @@ export class ProjectService extends BaseService {
         return results;
     }
 
+    async streamResultsToLocalFile(
+        callback: (writer: (data: ResultRow) => void) => Promise<void>,
+    ): Promise<string> {
+        const downloadFileId = nanoid(); // Creates a new nanoid for the download file because the jobId is already exposed
+        const filePath = `/tmp/${downloadFileId}.jsonl`;
+        await this.downloadFileModel.createDownloadFile(
+            downloadFileId,
+            filePath,
+            DownloadFileType.JSONL,
+        );
+        const writeStream = fs.createWriteStream(filePath, {
+            encoding: 'utf8',
+        });
+
+        writeStream.on('error', (err) => {
+            this.logger.error('Error writing to file', err);
+            throw new UnexpectedServerError('Error writing to file');
+        });
+
+        const writer = (data: ResultRow) => {
+            writeStream.write(`${JSON.stringify(data)}\n`);
+        };
+
+        try {
+            await callback(writer);
+        } catch (err) {
+            this.logger.error('Error during streaming', err);
+            throw err;
+        } finally {
+            writeStream.end(() => {
+                this.logger.debug('File has been saved.');
+            });
+        }
+
+        return downloadFileId;
+    }
+
+    async streamSqlQueryIntoFile(
+        userUuid: string,
+        projectUuid: string,
+        sql: string,
+    ): Promise<string> {
+        const { organizationUuid } = await this.projectModel.getSummary(
+            projectUuid,
+        );
+
+        await this.analytics.track({
+            userId: userUuid,
+            event: 'sql.executed',
+            properties: {
+                projectId: projectUuid,
+            },
+        });
+        const { warehouseClient, sshTunnel } = await this._getWarehouseClient(
+            projectUuid,
+            await this.getWarehouseCredentials(projectUuid, userUuid),
+        );
+        this.logger.debug(`Stream query against warehouse`);
+        const queryTags: RunQueryTags = {
+            organization_uuid: organizationUuid,
+            user_uuid: userUuid,
+        };
+
+        // TODO upload to s3 if enabled
+        const fileId = await this.streamResultsToLocalFile(async (writter) => {
+            await warehouseClient.streamQuery(
+                sql,
+                async ({ rows, fields }) => {
+                    const formattedRows = formatRows(rows, {}); // TODO fields to itemmap
```

Comment:
we should convert fields to itemmap to properly do formatting... I'm sure how to do it here (and efficiently)